lib/cn/cn_tree.c

```diff
@@ -2354,6 +2356,91 @@ cn_tree_fini(void)
     }
 }
 
+static int
+kblk_compare(const void *a, const void *b)
+{
+    const struct kvset_kblk *kblk_a = a, *kblk_b = b;
+
+    /* Return the kblock with the largest min key. */
+    return -1 * key_disc_cmp(&kblk_a->kb_kdisc_min, &kblk_b->kb_kdisc_min);
+}
+
+/**
+ *
+ */
+static merr_t
+get_split_key(const struct cn_tree_node *node, void *key_buf, size_t key_buf_sz, uint16_t *key_len)
+{
+    merr_t err;
+    struct bin_heap2 *bh;
+    struct kvset_list_entry *le;
+    struct kvset_kblk *kblk;
+    struct other_iterator **iters = NULL;
+    struct element_source **srcs = NULL;
+    uint64_t kvset_idx = 0;
+    uint64_t wlen = 0;
+    const uint64_t num_kvsets = cn_ns_kvsets(&node->tn_ns);
+    const uint64_t total_wlen = cn_ns_wlen(&node->tn_ns);
+
+    err = bin_heap2_create(num_kvsets, kblk_compare, &bh);
+    if (ev(err))
+        return err;
+
+    /* Forgive me for I have sinned; allocate all necessary memory for managing
+     * iterators and element sources in one go.
+     */
+    iters = calloc(1, sizeof(void *) * 2 * num_kvsets);
+    if (ev(!iters)) {
+        err = merr(ENOMEM);
+        goto out;
+    }
+
+    srcs = (struct element_source **)(iters + num_kvsets);
+
+    list_for_each_entry(le, &node->tn_kvset_list, le_link) {
+        struct other_iterator *iter = *(iters + kvset_idx);
+
+        err = other_iterator_create(le->le_kvset, &iter);
+        if (err)
+            goto out;
+
+        *(srcs + kvset_idx) = other_iterator_get_element_source(iter);
+
+        kvset_idx++;
+    }
+
+    assert(kvset_idx == num_kvsets);
+
+    err = bin_heap2_prepare(bh, num_kvsets, srcs);
+    if (ev(err))
+        goto out;
+
+    while (bin_heap2_pop(bh, (void **)&kblk)) {
+        wlen += kblk->kb_metrics.tot_key_bytes + kblk->kb_metrics.tot_val_bytes;
+        if (wlen >= total_wlen / 2)
+            break;
+    }
+
+    if (key_buf) {
+        const size_t copy_len = kblk->kb_klen_min < key_buf_sz ? kblk->kb_klen_min : key_buf_sz;
+
+        memcpy(key_buf, kblk->kb_koff_min, copy_len);
+    }
+    if (key_len)
+        *key_len = kblk->kb_klen_min;
+
+out:
+    if (iters) {
+        for (uint64_t i = 0; i < num_kvsets; i++)
+            other_iterator_destroy(*(iters + i));
+    }
```

Comment:
There's probably not enough RAM in the universe to support 2^32 kvsets (let along 2^64)!!!
Does cn_ns_kvsets() really return a 64-bit integer?

Reply:
Except that the bh is comprised of the iterators, so tearing down the iterators first followed by bh could touch freed memory, dunno, would have to look...

Reply:
yes, it does.

Reply:
Ok, maybe enough RAM to support 2^32, but definitely not 2^64 :-)